src/util/strencodings.cpp

```diff
@@ -46,14 +47,17 @@ bool IsHex(std::string_view str)
     return (str.size() > 0) && (str.size()%2 == 0);
 }
 
-bool IsHexNumber(std::string_view str)
+std::optional<std::string> TrySanitizeHexNumber(std::string_view input, int result_size)
 {
-    if (str.substr(0, 2) == "0x") str.remove_prefix(2);
-    for (char c : str) {
-        if (HexDigit(c) < 0) return false;
+    input = util::RemovePrefixView(input, "0x");
+    const auto final_size{(result_size < 0) ? input.size() : static_cast<size_t>(result_size)};
+    if (input.empty() || (input.size() > final_size)) return std::nullopt;
+    for (char c : input) {
+        if (HexDigit(c) < 0) return std::nullopt;
     }
-    // Return false for empty string or "0x".
-    return str.size() > 0;
+    std::string result(final_size, '0');
+    std::copy(input.begin(), input.end(), result.begin() + (final_size - input.size()));
+    return result;
 }
```

Comment:
Given that:
* this is only used in a single place in prod (which accept `std::string_view` as well and has an `std::string` input)
* we're always copying the result before returning
* the overwhelming majority of inputs will probably have the correct size already, i.e. don't need change.
* the performance of the method seems to be important

Could we either return an `std::optional<std::string_view>` instead or take an `std::string& input` to be able to return it?

i.e.
```suggestion
std::optional<std::string> TrySanitizeHexNumber(std::string input, int result_size)
{
    input = util::RemovePrefixView(input, "0x");
    const auto final_size{(result_size < 0) ? input.size() : static_cast<size_t>(result_size)};
    if (input.empty() || (input.size() > final_size)) return std::nullopt;
    for (char c : input) {
        if (HexDigit(c) < 0) [[unlikely]] return std::nullopt;
    }
    if (input.size() == final_size) [[likely]] return input;

    std::string result(final_size, '0');
    std::copy(input.begin(), input.end(), result.begin() + (final_size - input.size()));
    return result;
}
```

Reply:
I agree taking `input` as `std::string` works better with the return type.

Instead of introducing a 2nd string at the end, one could simply:
```C++
    input.insert(0, final_size - input.size(), '0');
    return input;
```

Reply:
>    input.insert(0, final_size - input.size(), '0');
    return input;

my functional programming past screams at the though of mutating parameters and also returning them
<insert meme with PTSD Chihuahua>

Reply:
Yeah, I had a bit of an itch as I was writing `return input`, but you made your bed when you started taking `std::string` by value. :)

Reply:
> the performance of the method seems to be important

How so? There is one callsite, which can be called at most once, to parse one optional debug-only option.

> Could we either return an std::optional<std::string_view>

How would that work? `std::string_view` is a non-owning view.

> we're always copying the result before returning

In 802374b4355bd1dec7a88bba6287c55f935699fe, we're always allocating exactly once. In your example, we'd be allocating at least once, and twice if we need padding. That doesn't look like an improvement to me.

> the overwhelming majority of inputs will probably have the correct size already, i.e. don't need change.

Given that it relies on user input, that seems hard to substantiate. In the functional tests, for example, we're using quite a bit of non-64 char hex input, which is reasonable given that we're doing all this work to accept number input.

Reply:
```C++
input = util::RemovePrefixView(input, "0x");
```
This line gives me the heebie-jeebies if input is `std::string`. It might be okay as we are assigning a shorter string, but worst case, the assignment operator could free it's prior heap buffer and allocate a shorter buffer, before attempting to copy from the string_view which now might reference the deallocated buffer. Using `RemovePrefix()` should alleviate the problem.

Reply:
Will let you decide, there are multiple ways to do this, each would be slightly better in a different scenario.
Whichever produces the cleanest code, choose that :)